cpp/src/join/join_kernels.cuh

```diff
@@ -0,0 +1,432 @@
+
+/* --------------------------------------------------------------------------*/
+/** 
+* @brief  Adds a pair of indices to the shared memory cache
+* 
+* @param[in] first The first index in the pair
+* @param[in] second The second index in the pair
+* @param[in,out] current_idx_shared Pointer to shared index that determines where in the shared
+memory cache the pair will be written
+* @param[in] warp_id The ID of the warp of the calling the thread
+* @param[out] joined_shared_l Pointer to the shared memory cache for left indices
+* @param[out] joined_shared_r Pointer to the shared memory cache for right indices
+* 
+*/
+/* ----------------------------------------------------------------------------*/
+template<typename size_type, typename output_index_type>
+__inline__ __device__ void add_pair_to_cache(const output_index_type first, 
+                                             const output_index_type second, 
+                                             size_type *current_idx_shared, 
+                                             const int warp_id, 
+                                             output_index_type *joined_shared_l,
+                                             output_index_type *joined_shared_r)
+{
+  size_type my_current_idx{atomicAdd(current_idx_shared + warp_id, size_type(1))};
+
+  // its guaranteed to fit into the shared cache
+  joined_shared_l[my_current_idx] = first;
+  joined_shared_r[my_current_idx] = second;
+}
+
+/* --------------------------------------------------------------------------*/
+/**
+* @brief  Builds a hash table from a table_device_view that maps the hash
+* values of each row to its respective row index.
+*
+* @param[in,out] multi_map The hash table to be built to insert rows into
+* @param[in] build_table The table to build the hash table on
+* @param[in] build_table_num_rows The number of rows in the build table
+* @tparam multimap_type The type of the hash table
+*
+*/
+/* ----------------------------------------------------------------------------*/
+template<typename multimap_type, typename ScalarDeviceViewType>
+__global__ void build_hash_table(multimap_type multi_map,
+                                 table_device_view build_table,
+                                 RowHash hash_build,
+                                 const cudf::size_type build_table_num_rows,
+                                 ScalarDeviceViewType error)
+{
+    cudf::size_type i = threadIdx.x + blockIdx.x * blockDim.x;
+
+    while (i < build_table_num_rows) {
+      // Compute the hash value of this row
+      const hash_value_type row_hash_value{hash_build(i)};
+
+      // Insert the (row hash value, row index) into the map
+      // using the row hash value to determine the location in the
+      // hash map where the new pair should be inserted
+      const auto insert_location = multi_map.insert(
+          thrust::make_pair(row_hash_value, i), true, row_hash_value);
```

Comment:
The `concurrent_unordered_multimap::insert` function should be redone away from this silliness that I did with passing in a `bool` for a precomputed hash value. It should just be two different overloads. The same is true for `find`.

Reply:
Jake can this be done separately? Suggest filing an issue.

Reply:
@jrhemstad ?

Reply:
Yes, this can be a different PR.